Add author profile preview cards to feed

aitchiss opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/main/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR is the final to address https://github.com/forem/rfcs/pull/222, adding the profile preview cards to posts in the feed.

The "Story" cards that are shown in the home feed are shown in other places too, and used for more than just Posts (e.g. users and comments in search results). In an effort to balance the consistency of user experience with the intentionally narrow scope of the RFC, I've implemented the profile preview cards for all "Story" cards for Posts. 

### Pages affected

- Home feed (regardless of time filter chosen)
- Tag index page
- Search results when "Posts" are selected
- A user or organization's profile page
- A series/collection page listing all articles in the series

When reviewing this PR it's worth bearing in mind that the feed is constructed in multiple places and for each page we'll need to check:

- Logged in vs logged out
- Initially displayed articles vs paged in articles

### Data loading strategy

A minimal version of the preview card is initially presented. This only includes:

- Name (link)
- Profile image
- Follow button

When you hover or focus on an author name, the remaining preview card content is fetched async, and then inserted into the preview card when it is ready. This means that you might notice (particularly on slower connections) the initial card loads only to expand vertically with more information. Similarly, you might also notice that the card border colour begins black, and then changes to the user's brand colour from their settings.

We only load an author's additional data _once_ per page. If an author has more than one article in the feed, we use an in-memory cache to populate any other preview cards of theirs that the user hovers/focuses.


## Related Tickets & Documents

https://github.com/forem/rfcs/pull/222

## QA Instructions, Screenshots, Recordings

For each of the areas mentioned above, we need to run through each of these checks _twice_ - once while logged out, and once while logged in.

**On mobile/small screens:**
-  the preview cards shouldn't show, and there should be no visible or functional change to the "story cards". This means that the author name should continue to be a link to the user's profile.
- Check this both as a logged in user and a logged out user

**On larger screens:**
- Hovering over an author's name shows a preview card. As noted above, this will initially show only minimal info, and then expand to show more detail. 
- You should notice in the dev tools network tab that if you now hover _another_ story card from the same author, that no additional network request is made to retrieve their details
- Clicking the author name (either by mouse or keyboard) opens the preview card as normal, and whether you use keyboard or mouse should have no impact on the loading of the additional data
- If you open the preview card using the keyboard, the user's name inside the preview card should be immediately focused
- Clicking the author name again, tabbing away, or pressing `Escape` should close the preview card. If you press `Escape` or click on something non-interactive (e.g. not a link or button) then the button which opened the card should be focused.
- Follow buttons within the preview card should behave as normal. This means that if you're logged out, they result in the login modal being shown, and if you're logged in, they accurately reflect the current "follow/following/follow back" state as you interact with them
- Check this both as a logged-in user and a logged out user

**Other things it would be useful to check**
- Story cards are also used for user and comment search results - we should double-check that they haven't been affected by this PR. This includes logged in vs out and large screen vs mobile.


### UI accessibility concerns?

I've included some accessibility testing steps above, but the main bulk of accessibility concerns for this RFC were addressed in the initial implementation and there _shouldn't_ be any regression here.

The main thing will be to make sure everything is operable by keyboard alone 🙂 

## Added/updated tests?

- [X] Yes - although this doesn't cover story cards that are constructed and displayed after paging in via infinite scroll. I'm not sure how best to approach that with Cypress


## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [X] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

N/A

## [optional] What gif best describes this PR or how it makes you feel?

![person dressed as a bear crosses a finish line](https://media.giphy.com/media/S797c9KnapZwWED2n4/giphy.gif)

rhymes commented on app/views/articles/_single_story.html.erb:
```diff
@@ -35,9 +35,16 @@
+              <button id="story-author-preview-trigger-<%= story.id %>" aria-controls="story-author-preview-content-<%= story.id %>" class="profile-preview-card__trigger fs-s px-0 crayons-btn crayons-btn--ghost p-0" aria-label="<%= story.cached_user.username %> profile details">
+                <%= story.cached_user.name %>
+              </button>
+              <%= render "/shared/profile_preview_card", actor: story.user, id: "story-author-preview-content-#{story.id}" %>
```
Unfortunately this partial is rendered by so many different pages that it's complicated to understand which `story` is referring to:

```shell
> rg "single_story"
app/views/collections/show.html.erb
18:          <%= render "articles/single_story", story: article.decorate, featured: article.main_image.present? %>

app/views/users/_main_feed.html.erb
10:      <%= render partial: "articles/single_story", collection: @pinned_stories, as: :story, locals: { featured: false } %>
15:  <%= render "articles/single_story", story: @stories.first, featured: false %>
19:  <%= render partial: "articles/single_story", collection: @stories[1, @stories.size], as: :story, locals: { featured: false } %>

app/views/stories/_main_stories_feed.html.erb
2:  <%= render partial: "articles/single_story", locals: { story: @pinned_article, pinned: true, featured: @pinned_article.id == @featured_story.id } %>
6:  <%= render partial: "articles/single_story", locals: { story: @featured_story, featured: true, pinned: false } %>
20:      <%= render partial: "articles/single_story", locals: { story: story, featured: false, pinned: false } %>

app/views/stories/tagged_articles/_main_feed.html.erb
1:<%= render partial: "articles/single_story", collection: @stories, as: :story, locals: { featured: false } %>

app/views/organizations/_main_feed.html.erb
2:<%= render partial: "articles/single_story", collection: @stories, as: :story, locals: { featured: false } %>
```

That said, since it's about the feed, it means that we need to go through all the places these files above are rendered from, find the corresponding controller action and load the `user` object, by adding `.includes([:user])` to the query.

I think `app/views/collections/show.html.erb` is fine since it's a `show` action (the error is present because its trying to render a list of objects, which incurs into a N + 1 query - each iteration loads the user object from the DB).

A place to start is `app/views/users/_main_feed.html.erb` which is rendered by `app/views/users/show.html.erb` which is rendered by https://github.com/forem/forem/blob/0ccb7608d0326d8320aea2303f7b280b87164458/app/controllers/stories_controller.rb#L213 which is basically the user profile page IIRC.

At some point in that method, the method `assign_user_stories` is called, which defines `@stories` and `@pinned_stories`:

https://github.com/forem/forem/blob/0ccb7608d0326d8320aea2303f7b280b87164458/app/controllers/stories_controller.rb#L311-L319

The statement `.includes([:user])` isn't present, which is why the test fails.

That said: which attributes of the `User` do you need, because IIRC the reason `_single_story` is using `cached_user` is for performance reasons and to avoid loading the entire user object in memory, maybe there's an alternative solution here where we add the required attributes to `.cached_user`...

aitchiss replied:
Thanks for all this detail - it's so helpful!

Yes, I was wondering if maybe the right solution would be to add to the cached user. As things stand at the moment, the attributes needed are:

- all the metadata that we show on the card (location, work, etc)
- the `user_colors` method which relies on `user.decorate.enriched_colors` and all that touches further down the line 😩 

It might be worth saying though that this is only affecting the **logged out** feed. Because that feed is initially generated server-side, I added the preview card in its entirety to the `html.erb` template. An alternative solution is that I handle the logged-out feed the same way as the logged in/paged in feed - i.e. show a minimal preview card (which I believe would only need the cached user details), and then fetch the remaining data view network call on hover.

I think that probably makes the most sense. I was trying to avoid unnecessary data fetches from the frontend, but stuffing loads of extra stuff in the cached user doesn't seem right for a feature a person might never use...

rhymes replied:
I like this last solution!

> I think that probably makes the most sense. I was trying to avoid unnecessary data fetches from the frontend, but stuffing loads of extra stuff in the cached user doesn't seem right for a feature a person might never use...

Also: the `cached_user` would need to be updated continuously and some of the data is not in the `users` table anymore, so it could be tricky

aitchiss replied:
Cool - sounds like a plan then! 😎  Thanks for your help!

nickytonline commented on app/javascript/packs/feedPreviewCards.jsx:
```diff
@@ -0,0 +1,109 @@
+
+async function populateMissingMetadata(metadataPlaceholder) {
+  const { authorId } = metadataPlaceholder.dataset;
+  const response = await fetch(`/profile_preview_cards/${authorId}`, {
```
Refactor (blocking): Use Forem's request utility.

https://github.com/forem/forem/blob/784201cdfc19fa0e7d4bdcc7f44b1da6594eef17/app/javascript/utilities/http/request.js#L1-L23

nickytonline commented on app/javascript/packs/feedPreviewCards.jsx:
```diff
@@ -0,0 +1,109 @@
+  );
+
+  for (const placeholder of allPlaceholdersForThisAuthor) {
+    render(
```
Refactor (potentially blocking): Can we render on mouseover instead? That way we only add it as we need it. I'm not positive if this will cause accessibility issues though. Regardless of that, this is a perfect use case for [memo](https://reactjs.org/docs/react-api.html#reactmemo) on the component.

nickytonline commented on app/javascript/packs/feedPreviewCards.jsx:
```diff
@@ -0,0 +1,109 @@
+      placeholder,
+    );
+
+    placeholder.parentElement.parentElement.style.setProperty(
```
Refactor (potentially blocking): If we know the selector for `placeholder.parentElement.parentElement`, prefer `placeholder.closest('some-container-selector')`

nickytonline commented on app/javascript/packs/feedPreviewCards.jsx:
```diff
@@ -0,0 +1,109 @@
+  const { target } = event;
+
+  if (target.classList.contains('profile-preview-card__trigger')) {
+    const metadataPlaceholder = target.parentElement.getElementsByClassName(
+      'author-preview-metadata-container',
+    )[0];
```
Refactor (blocking): Prefer Element.closest()

```suggestion
    const metadataPlaceholder = target.closest('author-preview-metadata-container');
```

aitchiss replied:
`closest` doesn't work in this context, because it will traverse up the DOM tree to find a match. Whereas, I need to find an element that lives further _down_ the DOM tree 🙁

nickytonline commented on app/javascript/packs/feedPreviewCards.jsx:
```diff
@@ -0,0 +1,109 @@
+}
+
+function listenForHoveredOrFocusedStoryCards() {
+  document
+    .getElementById('main-content')
+    .addEventListener('mouseover', checkForPreviewCardDetails);
+
+  document
+    .getElementById('main-content')
+    .addEventListener('focusin', checkForPreviewCardDetails);
```
Refactor (blocking): Store the element in a variable and reuse it.

```suggestion
  const mainContent = document
    .getElementById('main-content')
    
  mainContent.addEventListener('mouseover', checkForPreviewCardDetails);
  mainContent.addEventListener('focusin', checkForPreviewCardDetails);
```

nickytonline reviewed (commented):
@aitchiss, I've done an initial review. 😎

aitchiss commented:
I don't think the currently failing Cypress test is strictly related to my changes. I added a seeded article that's part of a series, and now the `pinArticle` test is failing with an eager loading error accessing the `series` attribute on the dashboard row:

```
2021-08-06 08:54:23 +0000 Rack app ("GET /dashboard" - (127.0.0.1)): #<Bullet::Notification::UnoptimizedQueryError: user: travis
GET /dashboard
USE eager loading detected
  Article => [:collection]
  Add to your query: .includes([:collection])
Call stack
  /home/travis/build/forem/forem/app/models/article.rb:436:in `series'
  /home/travis/build/forem/forem/app/views/dashboards/_dashboard_article_row.html.erb:25:in `_app_views_dashboards__dashboard_article_row_html_erb__2516742377942928274_1405720'
  /home/travis/build/forem/forem/app/views/dashboards/show.html.erb:46:in `block in _app_views_dashboards_show_html_erb__1983760403141576611_1405160'
  /home/travis/build/forem/forem/app/views/dashboards/show.html.erb:42:in `each'
  /home/travis/build/forem/forem/app/views/dashboards/show.html.erb:42:in `_app_views_dashboards_show_html_erb__1983760403141576611_1405160'
  /home/travis/build/forem/forem/app/lib/middlewares/set_time_zone.rb:10:in `call'
```

I'll have a further dig, but I think this is likely to be a separate fix.

aitchiss commented on spec/system/search/display_articles_search_spec.rb:
```diff
@@ -28,7 +28,6 @@
     expect(page).to have_selector("button[data-reactable-id=\"#{found_article_one.id}\"]")
     expect(page).to have_content("5 min read")
     expect(find_link("#ruby")["href"]).to include("/t/ruby")
-    expect(find_link(found_article_one.user.name)["href"]).to include(found_article_one.username)
```
I've removed this because the broader tests required for making sure the right preview details are added are now in Cypress

nickytonline commented on app/javascript/packs/feedPreviewCards.jsx:
```diff
@@ -0,0 +1,109 @@
+  observer.disconnect();
+});
+
+initializeFeedPreviewCards();
+listenForHoveredOrFocusedStoryCards();
```
Refactor (blocking): Use a [dynamic import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#dynamic_imports) since these preview cards are only for feed items that appear with infinite scrolling.

```suggestion
import('./someFile').then(
  ({ initializeFeedPreviewCards, listenForHoveredOrFocusedStoryCards }) => {
    initializeFeedPreviewCards();
    listenForHoveredOrFocusedStoryCards();
  },
);

```

aitchiss replied:
They're also needed for non-infinite-scroll feed preview cards that are present from the get-go. It's the mutation observer further up that picks up on newly inserted/paged in feed cards

nickytonline replied:
If they're used for non-infinite scrolling as well would it still be fine as a dynamic import? I'm on mobile ATM, but I imagine yes?

aitchiss replied:
I guess they could, but I'm not sure I understand what the benefit would be, since this pack is already loaded with `defer: true` and moving those functions to a separate file wouldn't leave anything for `feedPreviewCards` to do other than import them in - maybe I'm missing something here 🙃

nickytonline replied:
It is deferred, but my point was more about loading things dynamically that aren't required immediately for the page to load/work. It's true, the pack file would be doing pretty much the dynamic import and calling those functions, but that's OK. Typically, pack files, which are [webpack entry points](https://webpack.js.org/concepts/entry-points/), shouldn't have too much going on in them. Having said that, we have pack files with a bunch of stuff going on in them, lol.

aitchiss replied:
Yeah maybe that's my confusion 😂  I can split it out

aitchiss replied:
done 🚀  I'll keep this in mind for new packs too 😄

nickytonline commented on app/javascript/profilePreviewCards/MinimalProfilePreviewCard.jsx:
```diff
@@ -0,0 +1,60 @@
+      id={contentId}
+      class="profile-preview-card__content crayons-dropdown"
+      style="border-top: var(--su-7) solid var(--card-color);"
+      data-testid="profile-preview-card"
```
Question (non-blocking): Is it possible to get the profile preview card without a `data-testid` in an accessible manner?

aitchiss replied:
I think the best we could do is to `findByRole...` for the trigger button, and somehow dynamically pull out the `aria-controls` attribute ID then do `cy.get('#someId')` 🙁  

We only use the `data-testid` to scope our queries within the preview card though, and the tests tend to all check e.g. focus being transferred to the right element inside when it opens/closes so I feel like we're still testing that the button/dropdown are properly linked

nickytonline commented on app/javascript/profilePreviewCards/UserMetadata.jsx:
```diff
@@ -0,0 +1,76 @@
+import { h, Fragment } from 'preact';
+import { memo } from 'preact/compat';
+
+/**
+ * Component which renders the user metadata detail in a profile preview card.
+ *
+ * @param {string} email The user's email (if set to be publicly displayed)
+ * @param {string} location The user's location
+ * @param {string} created_at The user's join date string
+ * @param {string} education The user's education detail
+ * @param {string} work The user's work details
+ */
```
Documentation (blocking): These are properties of the props object. Intellisense/JSDoc will not work. Prefix all parameters with `props.`

```suggestion
/**
 * Component which renders the user metadata detail in a profile preview card.
 *
 * @param {object} props
 * @param {string} props.email The user's email (if set to be publicly displayed)
 * @param {string} props.location The user's location
 * @param {string} props.created_at The user's join date string
 * @param {string} props.education The user's education detail
 * @param {string} props.work The user's work details
 */
```

Before

![Incorrect JSDoc comments for the UserMetaData component](https://user-images.githubusercontent.com/833231/129055284-ef8f5853-f793-4918-888a-4f8a16aeeebb.png)

After

![Correct JSDoc comments for the UserMetaData component](https://user-images.githubusercontent.com/833231/129055163-44b82619-e8ba-457a-9625-607c91c35ae7.png)

nickytonline commented on app/javascript/profilePreviewCards/UserMetadata.jsx:
```diff
@@ -0,0 +1,76 @@
+  );
+};
+
+export const MemoizedUserMetadata = memo(UserMetadata);
```
Nitpick (blocking): Consider a name that does not include an implementation detail, or add the `memo` when the component it defined, e.g. `const UserMetadata = memo(({...}) => ...)`, since `UserMetaData` is not imported/used elsewhere.

nickytonline reviewed (changes requested):
Some request changes, but this is looking really good! 😎

aitchiss commented:
Thanks for the review @nickytonline! I've addressed the changes, and I've also added the new "reposition dropdowns on scroll" handler to this pack. I'm a bit wary about how the performance of this scroll handler is going to hold up as a user pages down through the feed infinite scroll though - do you think it'll be sufficient to test this once deployed to DEV? It would be very straightforward to rollback only the scroll listener changes if we needed to.

nickytonline reviewed (approved):
Thanks for making the changes! Let's gooo! 🚀
